compiler/src/iree/compiler/Codegen/Common/WorkgroupSpecializationPass.cpp

```diff
@@ -0,0 +1,276 @@
+    // instruction in the body.
+    auto minOp = llvm::dyn_cast_or_null<AffineMinOp>(block->front());
+    if (!minOp || minOp != minOp0) {
+      minOp0->moveBefore(&block->front());
+    }
+    builder.setInsertionPointAfter(&block->front());
+  } else {
+    builder.setInsertionPoint(&block->front());
+  }
+
+  // create a condition for scf.if
+  Value cond;
+  SmallVector<Value> constantOps;  // ConstantIndexOps for tile sizes
+  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
+    // clone the minSize op in the loop and place it before scf.if
+    AffineMinOp minOp = minSizeOps[i];
+    if (!minOp) {
+      // add a dummy value to indicate that this loop does not have an
+      // op for the bounded tile size.
+      constantOps.push_back(Value());
+      continue;
+    }
+
+    // Generate a compare op that checks the dynamic size is equal to the
+    // constant main tile size.
+    Value constant = builder.create<arith::ConstantIndexOp>(loc, tileSizes[i]);
+    constantOps.push_back(constant);
+    Value cmp = builder.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                              minOp, constant);
+    cond = cond ? builder.create<arith::AndIOp>(loc, cond, cmp) : cmp;
+  }
+
+  // generate scf.if %cond
+  auto ifOp = builder.create<scf::IfOp>(loc, cond, /*withElseRegion=*/true);
+
+  // Transfer the original body to the scf.else body.
+  auto origBodyBegin = ++Block::iterator(ifOp);
+  auto origBodyEnd = --block->end();  // yield
+
+  Block *elseBlock = ifOp.elseBlock();
+  elseBlock->getOperations().splice(elseBlock->begin(), block->getOperations(),
+                                    origBodyBegin, origBodyEnd);
+
+  auto b = ifOp.getThenBodyBuilder();
+  BlockAndValueMapping bvm;
+  for (unsigned i = 0, e = minSizeOps.size(); i != e; ++i) {
+    bvm.map(minSizeOps[i], constantOps[i]);
+  }
+  for (auto &blockOp : elseBlock->without_terminator()) {
+    b.clone(blockOp, bvm);
+  }
+  return;
+}
+
+namespace {
+struct WorkgroupSpecializationPass
+    : public WorkgroupSpecializationBase<WorkgroupSpecializationPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AffineDialect, IREE::HAL::HALDialect, linalg::LinalgDialect,
+                    scf::SCFDialect, tensor::TensorDialect>();
+  }
+
+  void runOnOperation() override {
+    if (!clEnableWorkgroupSpecialization) return;
+
+    IREE::HAL::ExecutableVariantOp variantOp = getOperation();
+    ModuleOp innerModule = variantOp.getInnerModule();
+    llvm::StringMap<IREE::HAL::ExecutableExportOp> entryPoints =
+        getAllEntryPoints(innerModule);
+
+    for (auto funcOp : innerModule.getOps<func::FuncOp>()) {
+      auto exportOp = entryPoints.lookup(funcOp.getName());
+      if (!exportOp) continue;
+
+      // Do not specialize when there is a scatter op since it
+      // may trigger share memory promotion for the output buffer
```

Comment:
This seems like a WAR for a bug downstream. Shouldnt promote a `flow.dispatch.tensor.store` if the source is a `scatter` op.... but I dont have a strong opinion.

Reply:
I removed this by taking Thomas' comment, which is to limit applying the pass for GPU vectorize, MatmulSimt, and transpose. Thanks